src/hotspot/share/runtime/sharedRuntime.cpp

```diff
@@ -1351,6 +1351,97 @@ bool SharedRuntime::resolve_sub_helper_internal(methodHandle callee_method, cons
   return true;
 }
 
+class Search2OopsClosure : public OopClosure {
+    bool _found;
+    const oop _o1;
+    const oop _o2;
+
+public:
+  Search2OopsClosure(oop obj1, oop obj2) : _found(false), _o1(obj1), _o2(obj2) { }
+
+  virtual void do_oop(oop* p) {
+    if (!_found) {
+      oop obj = NMethodAccess<ON_PHANTOM_OOP_REF | AS_NO_KEEPALIVE>::oop_load(p);
+      if (obj != nullptr && (obj == _o1 || obj == _o2)) {
+        _found = true;
+      }
+    }
+  }
+
+  virtual void do_oop(narrowOop* p) {
+    ShouldNotReachHere();
+  }
+
+  bool found() const {
+    return _found;
+  }
+};
+
+// Check if the holder of the target method is reachable from the caller nmethod if
+// the call is statically bound.
+// Without that path the target method could get unloaded even if the caller
+// nmethod is not unloading and a dangling Method* could be left int the static
+// stub for the call.
+// (See also ClassLoaderData::record_dependency())
+static void check_path_to_callee(bool is_virtual,
+                                 bool is_optimized,
+                                 CompiledMethod* from_cm,
+                                 Handle receiver,
+                                 methodHandle to_method,
+                                 Bytecodes::Code invoke_code,
+                                 JavaThread* thread) {
+  nmethod* from_nm = from_cm->as_nmethod_or_null();
+  InstanceKlass* to_klass = to_method->method_holder();
+  oop to_holder = to_klass->klass_holder();
+  ClassLoaderData * const to_cld = to_klass->class_loader_data();
+  if (from_nm != nullptr && !to_cld->is_permanent_class_loader_data()) {
+    RegisterMap regmap(thread,
+                       RegisterMap::UpdateMap::skip,
+                       RegisterMap::ProcessFrames::include,
+                       RegisterMap::WalkContinuation::skip);
+    Method* from_method = thread->last_java_vframe(&regmap)->method();
+    ClassLoaderData* const from_cld = from_method->method_holder()->class_loader_data();
+
+    if (to_cld->has_class_mirror_holder()) {
+      if (from_cld == to_cld) {
+        return;
+      }
+    } else {
+      oop to = to_cld->class_loader();
+      oop from = from_cld->class_loader();
+
+      if (from == to ||
+          (!from_cld->is_boot_class_loader_data() &&
+           java_lang_ClassLoader::isAncestor(from, to))) {
+        return; // `to` is reachable by iterating parents of `from`
+      }
+    }
```

Comment:
I'd be happier if this part was a function in ClassLoaderData or refactored from record_dependency (along with the constains function).  Since it's similar code.

Reply:
You are right. I've done the refactoring. Please let me know what you think.

Reply:
I like the refactoring.  Thank you.